src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -250,59 +253,85 @@ private void MarkEntireTypeInternal (TypeDefinition type, bool includeBaseTypes,
 		{
 #if DEBUG
 			if (!_entireTypeReasons.Contains (reason.Kind))
-				throw new ArgumentOutOfRangeException ($"Internal error: unsupported type dependency {reason.Kind}");
+				throw new InternalErrorException ($"Unsupported type dependency '{reason.Kind}'.");
 #endif
 
 			if (!typesAlreadyVisited.Add (type))
 				return;
 
 			if (type.HasNestedTypes) {
 				foreach (TypeDefinition nested in type.NestedTypes)
-					MarkEntireTypeInternal (nested, includeBaseTypes, new DependencyInfo (DependencyKind.NestedType, type), type, typesAlreadyVisited);
+					MarkEntireTypeInternal (nested, includeBaseTypes, GetDependencyInfo (nested, reason), type, typesAlreadyVisited);
 			}
 
 			Annotations.Mark (type, reason);
 			var baseTypeDefinition = type.BaseType?.Resolve ();
 			if (includeBaseTypes && baseTypeDefinition != null) {
-				MarkEntireTypeInternal (baseTypeDefinition, includeBaseTypes: true, new DependencyInfo (DependencyKind.BaseType, type), type, typesAlreadyVisited);
+				MarkEntireTypeInternal (baseTypeDefinition, includeBaseTypes: true, GetDependencyInfo (baseTypeDefinition, reason), type, typesAlreadyVisited);
 			}
-			MarkCustomAttributes (type, new DependencyInfo (DependencyKind.CustomAttribute, type), type);
+			MarkCustomAttributes (type, new DependencyInfo (reason.Kind == DependencyKind.AccessedViaReflection ? reason.Kind : DependencyKind.CustomAttribute, type), type);
 			MarkTypeSpecialCustomAttributes (type);
 
 			if (type.HasInterfaces) {
 				foreach (InterfaceImplementation iface in type.Interfaces) {
-					MarkInterfaceImplementation (iface, type);
+					if (reason.Kind == DependencyKind.AccessedViaReflection)
```

Comment:
I would prefer to pass this difference in as a separate parameter. It would make it cleaner in my mind. Just like we have a parameter which dictates if base types should be included or not, we should add one which dictates if interfaces should be fully included or not.